pageserver/src/tenant.rs

```diff
@@ -1519,30 +1537,11 @@ impl Tenant {
             .context("resume_deletion")
             .map_err(LoadLocalTimelineError::ResumeDeletion)?;
         }
-        // Complete deletions for offloaded timeline id's.
-        offloaded_timelines_list
-            .retain(|(offloaded_id, _offloaded)| {
-                // At this point, offloaded_timeline_ids has the list of all offloaded timelines
-                // without a prefix in S3, so they are inexistent.
-                // In the end, existence of a timeline is finally determined by the existence of an index-part.json in remote storage.
-                // If there is a dangling reference in another location, they need to be cleaned up.
-                let delete = offloaded_timeline_ids.contains(offloaded_id);
-                if delete {
-                    tracing::info!("Removing offloaded timeline {offloaded_id} from manifest as no remote prefix was found");
-                }
-                !delete
-        });
-        if !offloaded_timelines_list.is_empty() {
-            tracing::info!(
-                "Tenant has {} offloaded timelines",
-                offloaded_timelines_list.len()
-            );
-        }
         {
             let mut offloaded_timelines_accessor = self.timelines_offloaded.lock().unwrap();
             offloaded_timelines_accessor.extend(offloaded_timelines_list.into_iter());
         }
-        if !offloaded_timeline_ids.is_empty() {
+        if !preload.tenant_manifest.offloaded_timelines.is_empty() {
```

Comment:
The

```
  offloaded_timelines_list
            .retain(|(offloaded_id, _offloaded)| {
```

earlier in this function does not update `preload.tenant_manifest.offloaded_timelines`.

Does it?

Reply:
Fair point, actually what my PR suggests right now would upload the manifest more or less unconditionally if there is any offloaded timelines in the manifest.

We can't use emptiness of `offloaded_timelines_list` as criterion here either, because we remove prefix-less timelines from that list. In the case we remove all offloaded timelines, `offladed_timelines_list` is empty, and we still want to write the manifest (otherwise it would still contain the deleted timelines).

I suppose I could add a `any_removed` bool, that is set to true if any item is removed from `offloaded_timelines_list`.

This might still lead us to sometimes have large generation gaps, because we only write the manifest if there is changes, but I think that's fair.

Reply:
I kinda don't like that we have have this `offloaded_timeline_ids` and `offloaded_timeline_list` and `preload.tenant_manifest.offloaded_timelines`.

Not immediately clear when reading whether it's required to keep the entries in sync or why we need three collections of that in the first place.

Not asking to fix it in this PR - it's just the first time I have taken a look at that code.

Reply:
`offloaded_timeline_ids` is a hash set to allow for low-cost membership tests. `offloaded_timeline_list` allows us to keep the order of the timelines. otherwise the order would be different each time we create a manifest.

`offloaded_timeline_list` is a vector of `OffloadedTimeline` objects, while `offloaded_timelines` contains `OffloadedTimelineManifest` objects.

Maybe a refactor could make `OffloadedTimeline` contain the `OffloadedTimelineManifest`, reducing duplication a little. On the other hand, I intend `OffloadedTimelineManifest` to be closer to the json while `OffloadedTimeline` should be what the internal representation of things is. IDK.

Reply:
> offloaded_timeline_ids is a hash set to allow for low-cost membership tests. offloaded_timeline_list allows us to keep the order of the timelines. otherwise the order would be different each time we create a manifest.

Yeah, not debating that.

My usual strategy for dealing with this sort of stuff is to create a wrapper collection

```
struct OffloadedTimelines {
  by_id: HashSet<TimelineId>,
  in_storage_order: Vec<...>,
}
```

Reply:
> actually what my PR suggests right now would upload the manifest more or less unconditionally if there is any offloaded timelines in the manifest.

note to self: this is still not resolved